### PR TITLE
add token support for OAUTH2

### PR DESCRIPTION
**Pull request by hmeiland**
**Description of the Pull Request (PR):**

PR for issue #1622 , add token support for docker registries with OAUTH2
tested with Azure Container Registry.

**This fixes or addresses the following GitHub issues:**

- Ref: #1622 

**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [ ] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware] (https://www.github.com/singularityware/singularityware.github.io) documentation base. -> to be done
- [x] I have tested this PR locally with a `make test`
- [x] This PR is NOT against the project's master branch
- [ ] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md) -> not my code, see real contributer in #1622
- [x] This PR is ready for review and/or merge

Attn: @singularityware-admin

### CODE REVIEW

**Comment by hmeiland** [comment]
please verify 248:80; not sure if we can change that without creating other best-practise issues...

File: libexec/python/docker/api.py [hunk 1]
```diff
@@ -225,7 +225,11 @@ def update_token(self, response=None, auth=None, expires_in=300):
 
         try:
             data = json.loads(response)
-            token = {"Authorization": "Bearer %s" % data["token"]}
+            if 'access_token' in data:
```
Comment (GodloveD) [inline libexec/python/docker/api.py:228]: OK, so I'm going to show some of my ignorance about this stuff.  Please bear with me.  

In the old version we always default to `data["token"]`.  In this PR, we default to `data["access_token"]` if it exists and only fall back to `data["token"]` in the event that it doesn't.  Cool.  

Now here is where my ignorance comes in.  Will there ever be a situation in which: 
1) `data["access_token"]` exists and 
2) it is *not* the correct token to use

In other words could both `access_token` and `token` exist in the same hash table in a situation where OAUTH2 should not really be used.  And then would the presence of `access_token` block the correct usage of `token`.

If this might be the case, maybe you should add a `try`.  If `access_token` fails fall back to `token` and try again.  

Is this just nutty?

**Review by GodloveD — CHANGES_REQUESTED** [review]
Can you please add this to the changelog under the enhancements section of 2.6?

Also my preference would be that both you and @hakonenger add yourselves to the authors file.

File: libexec/python/docker/api.py (hunk 1 above)
Comment (dtrudg) [reply libexec/python/docker/api.py:228]: This code will be fine as if `access_token` is present you are using the OAUTH2 path, which doesn't set a `token`. `token` will appear alone if you use the docker token authentication path.

Edit.... actually if an auth service is returning both (which would be very odd as is out of spec for the 2 types) we are still good by the client spec...

> At least one of these fields must be specified, but both may also appear (for compatibility with older clients). When both are specified, they should be equivalent; if they differ the client's choice is undefined.

**Review by dtrudg — APPROVED** [review]
This looks good to me. I'm not sure that it will handle all eventualities with the OAUTH2 auth flow. We don't handle e.g. refresh tokens at all, but if this allows basic usage of e.g. Azure I'm pretty confident it won't cause regression elsewhere.

**Comment by dtrudg** [comment]
@hmeiland - for the PEP8 warning can you put a `# noqa` comment on the end of the line. I think this is one where making the linter happy is going to result in something less readable, which is not what we want.

File: libexec/python/docker/api.py (hunk 1 above)
Comment (GodloveD) [reply libexec/python/docker/api.py:228]: Thanks for the clarification @dctrud !

**Comment by GodloveD** [comment]
So, just to summarize, add to changelog, add the `# noqa`, and add to contributors (optional) and I'll approve.  Thanks!

**Comment by hmeiland** [comment]
Added # noqa, added contributers and added to changelog -> made bugfix entry for upcoming 2.6 release
Let me know if this ticks all the boxes...

**Review by GodloveD — APPROVED** [review]
lgtm

**Comment by GodloveD** [comment]
Thanks @hmeiland and @dctrud !
